File: src/DashSpec.Core/Parsing/ToolbarPlacementResolver.cs
using DashSpec.Core.Layout;
using DashSpec.Core.Model;
using DashSpec.Core.Parsing;

namespace DashSpec.Core.Parsing;

internal static class ToolbarPlacementResolver
{
    public static IReadOnlyList<string> ResolveFilterNames(
        IReadOnlyList<FilterDefinition> filters,
        IReadOnlyList<string> flatNames,
        LayoutBoardDefinition? board)
    {
        if (board is null)
        {
            return flatNames;
        }

        if (flatNames.Count > 0)
        {
            throw new DashSpecParseException(
                "Toolbar cannot combine a layout board with a flat filter list.");
        }

        const string context = "Toolbar";
        var names = new List<string>();
        foreach (var row in board.Rows)
        {
            foreach (var token in row)
            {
                var name = FilterLayoutRefResolver.Resolve(token, filters, context);
                if (names.Contains(name, StringComparer.OrdinalIgnoreCase))
                {
                    throw new DashSpecParseException(
                        $"{context}: filter '{name}' appears more than once in the toolbar board.");
                }

                names.Add(name);
            }
        }

        if (names.Count == 0)
        {
            throw new DashSpecParseException($"{context} layout board requires at least one filter.");
        }

        return names;
    }
}

File: src/DashSpec.Modeling.Parse/Document/ToolbarPlacementResolver.fs
namespace DashSpec.Modeling.Parse.Document

open System
open System.Collections.Generic
open DashSpec.Modeling.Core
open DashSpec.Modeling.Parse.Filter
open DashSpec.Modeling.Parse.Layout

/// Simplified resolver — full FilterLayoutRefResolver pending port.
module ToolbarPlacementResolver =

    let resolveFilterNames
        (filters: IReadOnlyList<FilterDefinition>)
        (flatNames: IReadOnlyList<string>)
        (board: LayoutBoardDefinition option)
        =
        match board with
        | None -> flatNames
        | Some b ->
            if flatNames.Count > 0 then
                raise (DashSpecParseException("Toolbar cannot combine a layout board with a flat filter list."))

            let names = ResizeArray<string>()

            for row in b.Rows do
                for token in row do
                    let name =
                        filters
                        |> Seq.tryFind (fun f ->
                            String.Equals(f.Name, token, StringComparison.OrdinalIgnoreCase)
                            || (f.LayoutRef.IsSome
                                && String.Equals(f.LayoutRef.Value, token, StringComparison.OrdinalIgnoreCase)))
                        |> function
                            | Some f -> f.Name
                            | None -> token

                    if names |> Seq.exists (fun n -> String.Equals(n, name, StringComparison.OrdinalIgnoreCase)) then
                        raise (DashSpecParseException($"Toolbar: filter '{name}' appears more than once in the toolbar board."))

                    names.Add name

            if names.Count = 0 then
                raise (DashSpecParseException("Toolbar layout board requires at least one filter."))

            names :> IReadOnlyList<_>
